### ISSUE TITLE
linter: no-control-regex should consider `\0` as valid

### ISSUE THREAD

**Issue by BPScott** [oxc-project/oxc#25024]
### What version of Oxlint are you using?

1.74.0

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{"rules": {"no-control-regex": "error"}}
```

### What happened?

eslint's `no-control-regex` rule considers `\0` to be an allowable escape code (like `\n`, `\t` and `\r`.
While oxlint's `no-control-regex` marks it as invalid.

Code to reproduce:

```
/\0\t\n\r/;
```

Expected behavour: no lint error in oxlint - matching what eslint does
Actual behaviour: lint error on `\0`

[oxlint playground](https://playground.oxc.rs/?lintRules=eslint%2Fno-control-regex&options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22tsx%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=%2F%5C0%5Ct%5Cn%5Cr%2F%3B)
[eslint playground](https://eslint.org/play/#eyJ0ZXh0IjoiL1xcMFxcdFxcblxcci87Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoic2NyaXB0IiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==)

---

It looks like this was raised last year in #10950, but the "Fix" PR for it retained \0 being considered an error, so it failed to retain parity with eslint's behavior (so to me it seems like the wrong fix.

**Comment by camc314** [comment]
I believe this is an intended behaviour `\0` is a control character which is why we report on it.

Does a disable comment not make sense for this use case? Do you have a use case to allow these chars?

**Comment by BPScott** [comment]
> I believe this is an intended behaviour \0 is a control character which is why we report on it. 

My interpretation was that `\0` falls into the same conceptual class as `\n`, `\r` and `\t` - they're single character escapes rather than codepoint references - all three of which both eslint and oxlint allow. See [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape#0) having \0 under a separate heading to the other "reference a codepoint with a number" syntaxes. To me that suggests that the current behaviour is inconsistent. 

Instead of writing `\x0A`, `\u000A` and `\u{A}` you should write `\n` (which is allowed).
Instead of writing `\x09`, `\u0009` and `\u{9}` you should write `\t` (which is allowed).
So I'd expect that instead of writing `\x00`, `\u0000` and `\u{0}` you should write `\0`.


If you're thinking `\0` should raise a lint error, then does it follow that `\n`, `\r`, `\t` should also raise a lint error?


> Does a disable comment not make sense for this use case? Do you have a use case to allow these chars?

I am migrating from eslint to oxlint and found code that passed in eslint but raised an error in oxlint. I raised this issue as I assumed that oxlint's intent was to implement the same behaviour as eslint, and this is a case where the behaviour of the rule diverges.  This isn't a critical problem, and I can indeed use a disable comment if you consider this divergence as acceptable.

Thanks for all your amazing work on the oxc ecosystem.

**Comment by camc314** [comment]
> > I believe this is an intended behaviour \0 is a control character which is why we report on it. 
> 
> My interpretation was that `\0` falls into the same conceptual class as `\n`, `\r` and `\t` - they're single character escapes rather than codepoint references - all three of which both eslint and oxlint allow. See [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape#0) having \0 under a separate heading to the other "reference a codepoint with a number" syntaxes. To me that suggests that the current behaviour is inconsistent. 
> 
> Instead of writing `\x0A`, `\u000A` and `\u{A}` you should write `\n` (which is allowed).
> Instead of writing `\x09`, `\u0009` and `\u{9}` you should write `\t` (which is allowed).
> So I'd expect that instead of writing `\x00`, `\u0000` and `\u{0}` you should write `\0`.
> 
> 
> If you're thinking `\0` should raise a lint error, then does it follow that `\n`, `\r`, `\t` should also raise a lint error?

this is incredibly well explained and justified. thanks for the clarity.

ill get this fixed shortly.
